core/network/fabric-interop-cc/contracts/interop/manage_assets.go

```diff
@@ -0,0 +1,304 @@
+	hasher.Write([]byte(preimage))
+	shaHash := hasher.Sum(nil)
+	//shaHashBase64 := base64.URLEncoding.EncodeToString(shaHash)
+	shaHashBase64 := base64.StdEncoding.EncodeToString(shaHash)
+
+	if shaHashBase64 == hashBase64 {
+		log.Info(fmt.Printf("checkIfCorrectPreimage: preimage %s is passed correctly.\n", preimage))
+	} else {
+		log.Info(fmt.Printf("checkIfCorrectPreimage: preimage %s is not passed correctly.\n", preimage))
+		return false, nil
+	}
+	return true, nil
+}
+
+// ClaimAsset cc is used to record claim of an asset on the ledger
+func (s *SmartContract) ClaimAsset(ctx contractapi.TransactionContextInterface, assetAgreementBytes string, claimInfoBytes string) error {
+
+	assetAgreement := &common.AssetExchangeAgreement{}
+	err := proto.Unmarshal([]byte(assetAgreementBytes), assetAgreement)
+	if err != nil {
+		log.Error(err.Error())
+		return err
+	}
+	// display the requested asset agreement
+	log.Info(fmt.Printf("assetExchangeAgreement: %+v\n", assetAgreement))
+
+	claimInfo := &common.AssetClaimHTLC{}
+	err = proto.Unmarshal([]byte(claimInfoBytes), claimInfo)
+	if err != nil {
+		errorMsg := fmt.Sprintf("Unmarshal error: %s", err)
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	// display the claim information
+	log.Info(fmt.Printf("claimInfo: %+v\n", claimInfo))
+
+	assetLockKey := assetAgreement.Type + ":" + assetAgreement.Id
+
+	assetLockValBytes, err := ctx.GetStub().GetState(assetLockKey)
+	if err != nil {
+		log.Error(err.Error())
+		return err
+	}
+
+        if assetLockValBytes == nil {
+		errorMsg := fmt.Sprintf("No asset of type %s and ID %s is locked", assetAgreement.Type, assetAgreement.Id)
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	assetLockVal := AssetLockValue{}
+	err = json.Unmarshal(assetLockValBytes, &assetLockVal)
+	if err != nil {
+		errorMsg := fmt.Sprintf("Unmarshal error: %s", err)
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	if assetLockVal.Locker != assetAgreement.Locker || assetLockVal.Recipient != assetAgreement.Recipient {
+		errorMsg := fmt.Sprintf("Can not claim asset of type %s and ID %s as it is locked by %s for %s", assetAgreement.Type, assetAgreement.Id, assetLockVal.Locker, assetLockVal.Recipient)
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	// compute the hash from the preimage
+	isCorrectPreimage, err := checkIfCorrectPreimage(string(claimInfo.HashPreimage), string(assetLockVal.Hash))
+	if err != nil {
+		errorMsg := fmt.Sprintf("Claim asset of type %s and ID %s error: %v", assetAgreement.Type, assetAgreement.Id, err)
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	if isCorrectPreimage == false {
+		errorMsg := fmt.Sprintf("Can not claim asset of type %s and ID %s as the hash preimage is not matching", assetAgreement.Type, assetAgreement.Id)
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	// Check if expiry time is elapsed
+	currentTimeSecs := time.Now().UnixNano() / int64(time.Second)
```

Comment:
Change `UnixNano()` to `Unix()`.

Also, from an efficiency perspective, I'd move this check above (i.e., before) the hash preimage check. If the time's expired, we'll save some computation cycles that would otherwise be spent in hashing.

Reply:
Done @VRamakrishna .